File: frontend/de/unika/ipd/grgen/ir/stmt/SwitchStatement.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ir.stmt;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;

import de.unika.ipd.grgen.ir.NeededEntities;
import de.unika.ipd.grgen.ir.expr.Expression;

/**
 * Represents a switch statement in the IR.
 */
public class SwitchStatement extends EvalStatement
{
	private Expression switchExpr;
	private Collection<CaseStatement> statements = new ArrayList<CaseStatement>();

	public SwitchStatement(Expression switchExpr)
	{
		super("switch statement");
		this.switchExpr = switchExpr;
	}

	public void addStatement(CaseStatement statement)
	{
		statements.add(statement);
	}

	public Expression getSwitchExpr()
	{
		return switchExpr;
	}

	public Collection<CaseStatement> getStatements()
	{
		return Collections.unmodifiableCollection(statements);
	}

	@Override
	public void collectNeededEntities(NeededEntities needs)
	{
		switchExpr.collectNeededEntities(needs);
		for(EvalStatement statement : statements) {
			statement.collectNeededEntities(needs);
		}
	}
}

File: frontend/de/unika/ipd/grgen/ir/stmt/SwitchStatement.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ir.stmt
{

	using System.Collections.Generic;

	using NeededEntities = de.unika.ipd.grgen.ir.NeededEntities;
	using Expression = de.unika.ipd.grgen.ir.expr.Expression;

	/// <summary>
	/// Represents a switch statement in the IR.
	/// </summary>
	public class SwitchStatement : EvalStatement
	{
		private Expression switchExpr;
		private List<CaseStatement> statements = new List<CaseStatement>();

		public SwitchStatement(Expression switchExpr)
			: base("switch statement")
		{
			this.switchExpr = switchExpr;
		}

		public virtual void AddStatement(CaseStatement statement)
		{
			statements.Add(statement);
		}

		public virtual Expression SwitchExpr
		{
			get
			{
				return switchExpr;
			}
		}

		public virtual ICollection<CaseStatement> Statements
		{
			get
			{
				return statements.AsReadOnly();
			}
		}

		public override void CollectNeededEntities(NeededEntities needs)
		{
			switchExpr.CollectNeededEntities(needs);
			foreach(EvalStatement statement in statements)
				statement.CollectNeededEntities(needs);
		}
	}

}
